File: ComputerScience/C/exercices/L2_C17/linked_list/gentil.pl


extends= /ComputerScience/C/template/std_progC17_clang.pl

@ gentil.py



title = Des histoires de mots

text==
# Mots Méchants Mots Gentils.

Le Père Noël a besoin d'aide pour déterminer quelles chaînes de son fichier texte sont méchantes ou gentilles.

Une chaîne gentille est une chaîne avec toutes les propriétés suivantes :

     Il contient au moins trois voyelles (aeiou uniquement), comme aei, xazegov ou aeiouaeiouaeiou.
     Il contient au moins une lettre qui apparait deux fois de suite, comme xx, abcdde (dd) ou aabbccdd (aa, bb, cc ou dd).
     Il ne contient pas les chaînes ab, cd, pq ou xy, même si elles font partie de l'une des autres exigences.

Par example:

     ugknbfddgicrmopn est agréable car il a au moins trois voyelles (u...i...o...), une lettre double (...dd...), et aucune des sous-chaînes interdites.
     aaa est agréable car il a au moins trois voyelles et une lettre double, même si les lettres utilisées par différentes règles se chevauchent.
     jchzalrnumimmnmhp est méchant car il n`a pas de double lettre.
     haegwjzuvuyypxyu est méchant car il contient la chaîne xy.
     dvszwmarrgswjxmb est méchant car il ne contient qu`une seule voyelle. 

Ecrire une fonction question1(char *filename) qui lit dans le fichier filename et
 compte et retourne le nombre de chaines (une par ligne dans le fichier) gentilles.


Le fichier "exemple" que vous pouvez copier si dessous contient 1000 chaines et le nombre de mots gentils est 253.





<div style="font-size:0.5em">
<details><summary>Fichier exemple <button onclick="docopy('exemple')">Copy exemple</button></summary><blockquote id="exemple" style="white-space: pre-line" 
>{{lexemple}}
</blockquote></details>

==


#@ gentil.test [exemple]
lexemple=@gentil.test


before==

# produire le fichier data 
import gentil

gentil.buildfile("data")
# calculer la solution pour ce fichier data
sol1 = gentil.Question1()
sol2 = gentil.Question2()

# crer un fichier 
with open("sol1.c","w") as f:
    f.write("""
    int solution(char *name){ 
    if (strcmp(name,"error") == 0 ) 
    return""")
    f.write(str(sol1))
    f.write(""";
    else return 253; 
    }""")

solution= """
    int solution(char *name){ 
    if (strcmp(name,"error") == 0 ) 
    return"""+str(sol1)+"""; else return 253; }"""

import subprocess 

def makeO(filename):
    """
    compile the source in argument and return 
    """
    command_args = ["clang", filename, "-c"] 
    sp = subprocess.run(command_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    spout = sp.stdout.decode()
    errout = sp.stderr.decode()
    returncode = sp.returncode
    return (returncode, spout, errout)

c= makeO("sol1.c")
if not c[0]:
    import sys
    print(c,file=sys.stderr)

# compilation et création du fichier sol1.o
# effasser le fichier sol1.c 
import os
os.remove("sol1.c")

# attention il faut ajouter sol1.o dans les "flags" de compilation du C 
 
# même chose pour le deuxième exercice avec sol2
==



code_before==

==

form= {{ editor | component }}


editor.code==
int Question1(){
// votre code 

}
==

code_after==
int main(){

printf("J'appel votre fonction : Question1() \n");
int rep =Question1();
int sol= solution()
if (rep > sol) printf(" Votre réponse est trop grande \n");
else if (rep < sol) printf(" Votre réponse est trop grande \n");
else printf("Bravo bonne réponse\n);

}
==

File: ComputerScience/C/exercices/L2_C17/linked_list/gentil.py




import os
import random


def randomletter():
    return chr(random.randint(97,97+25))

def randomword(longueur):
    l=[]
    for i in range(longueur):
        l.append(randomletter())
    return "".join(l)


def buildfile(filename):
    """Build
    """
    with open(filename, 'w') as f:
        for i in range(1000):
            print(randomword(16),file=f)
    

def doit():
    """
    Crer le fichier "data" et retourne deux pltests
    """
    # Creation du fichier "data" 
    buildfile("data")
    # Cration des pltest retourne deux pltests
    sol1 = Question1()
    sol2 = Question2()
    
    

    pltest= f"""
    >>> question1() == {sol1} # Question 1 étage 
    True
    """
    pltest2= f"""
    >>> question2() == {sol2} # Question 2 position 
    True
    """
    return pltest,pltest2


from re import A

voielles = "aeiou"

def Question1():
    with open("data","r") as f:
        nice= 0
        for line in f.readlines():
            line = line.strip()
            nbv=0
            dwble=0
            taboo=True
            last=""
            for x in line:
                if x in voielles:
                    nbv+=1
                if x == last:
                    dwble+=1
                if last+x in ["ab", "cd", "pq" , "xy"]:
                    taboo = False
                last = x
            if taboo and nbv >= 3 and dwble >= 1 :
                nice = nice +1 

    return nice        


def pairpresent(i,s):
    """
    je cherche si s[i]s[i+1] réapparait plus loin dans la liste.
    """
    j=i+2
    while j <len(s)-1 :
        if s[i]==s[j] and s[i+1]==s[j+1]:
            return True
        j = j + 1
    return False
    
    

def nice(s):
    """
        qjhvhtzxzqqjkmpb est sympa car il a une paire qui apparaît deux fois (qj) et une lettre qui se répète avec exactement une lettre entre elles (zxz).
    xxyxx est agréable car il a une paire qui apparaît deux fois et une lettre qui se répète avec une entre, même si les lettres utilisées par chaque règle se chevauchent.
    uurcxstgmygtbstg est méchant car il a une paire (tg) mais pas de répétition avec une seule lettre entre eux.
    ieodomkazucvgmuy est méchant car il a une lettre répétitive avec une entre (odo), mais aucune paire qui apparaît deux fois.
    >>> nice("qjhvhtzxzqqjkmpb")
    True
    >>> nice("xxyxx")
    True
    >>> nice("uurcxstgmygtbstg")
    False
    >>> nice("ieodomkazucvgmuy")
    False
    """
    pair=False
    triplet = False
    for i in range(len(s)-2):
        pair = pair or pairpresent(i,s)
        if s[i]==s[i+2] :
            triplet=True
        if pair and triplet:
            return True
    return False

def Question2():
    with open("data","r") as f:
        nic= 0
        for line in f.readlines():
            line = line.strip()
            if nice(line):
                nic+=1
        return nic

print(doit())
